External/pubchem_shape/PubChemShape.cpp

```diff
@@ -0,0 +1,424 @@
+    {37, 3.03},  // Rb
+    {38, 2.49},  // Sr
+    {49, 1.93},  // In
+    {50, 2.17},  // Sn
+    {51, 2.06},  // Sb
+    {52, 2.06},  // Te
+    {53, 1.98},  // I
+    {54, 2.16},  // Xe
+    {55, 3.43},  // Cs
+    {56, 2.68},  // Ba
+    {81, 1.96},  // Tl
+    {82, 2.02},  // Pb
+    {83, 2.07},  // Bi
+    {84, 1.97},  // Po
+    {85, 2.02},  // At
+    {86, 2.20},  // Rn
+    {87, 3.48},  // Fr
+    {88, 2.83},  // Ra
+};
+constexpr double radius_color =
+    1.08265;  // same radius for all feature/color "atoms"
+
+namespace {
+class ss_matcher {
+ public:
+  ss_matcher(const std::string &pattern) : m_pattern(pattern) {
+    m_needCopies = (pattern.find_first_of("$") != std::string::npos);
+    RDKit::RWMol *p = RDKit::SmartsToMol(pattern);
+    m_matcher = p;
+    POSTCONDITION(m_matcher, "no matcher");
+  };
+  const RDKit::ROMol *getMatcher() const { return m_matcher; };
+  unsigned int countMatches(const RDKit::ROMol &mol) const {
+    PRECONDITION(m_matcher, "no matcher");
+    std::vector<RDKit::MatchVectType> matches;
+    // This is an ugly one. Recursive queries aren't thread safe.
+    // Unfortunately we have to take a performance hit here in order
+    // to guarantee thread safety
+    if (m_needCopies) {
+      const RDKit::ROMol nm(*(m_matcher), true);
+      RDKit::SubstructMatch(mol, nm, matches);
+    } else {
+      const RDKit::ROMol &nm = *m_matcher;
+      RDKit::SubstructMatch(mol, nm, matches);
+    }
+    return matches.size();
+  }
+  ~ss_matcher() { delete m_matcher; };
+
+ private:
+  ss_matcher() : m_pattern(""){};
+  std::string m_pattern;
+  bool m_needCopies{false};
+  const RDKit::ROMol *m_matcher{nullptr};
+};
+}  // namespace
+
+typedef boost::flyweight<boost::flyweights::key_value<std::string, ss_matcher>,
+                         boost::flyweights::no_tracking>
+    pattern_flyweight;
+// Definitions for feature points adapted from:
+// Gobbi and Poppinger, Biotech. Bioeng. _61_ 47-54 (1998)
+const std::vector<std::vector<std::string>> smartsPatterns = {
+    {"[$([N;!H0;v3,v4&+1]),\
+$([O,S;H1;+0]),\
+n&H1&+0]"},                                // Donor
+    {"[$([O,S;H1;v2;!$(*-*=[O,N,P,S])]),\
+$([O,S;H0;v2]),\
+$([O,S;-]),\
+$([N;v3;!$(N-*=[O,N,P,S])]),\
+n&H0&+0,\
+$([o,s;+0;!$([o,s]:n);!$([o,s]:c:n)])]"},  // Acceptor
+    {
+        "[r]1[r][r]1",
+        "[r]1[r][r][r]1",
+        "[r]1[r][r][r][r]1",
+        "[r]1[r][r][r][r][r]1",
+        "[r]1[r][r][r][r][r][r]1",
+    },  // rings
+        //    "[a]",                                                  //
+        //    Aromatic
+        //    "[F,Cl,Br,I]",                                          // Halogen
+    {"[#7;+,\
+$([N;H2&+0][$([C,a]);!$([C,a](=O))]),\
+$([N;H1&+0]([$([C,a]);!$([C,a](=O))])[$([C,a]);!$([C,a](=O))]),\
+$([N;H0&+0]([C;!$(C(=O))])([C;!$(C(=O))])[C;!$(C(=O))])]"},  // Basic
+    {"[$([C,S](=[O,S,P])-[O;H1,-1])]"}                       // Acidic
+};
+std::vector<std::vector<const ROMol *>> *getPh4Patterns() {
```

Comment:
This could return

```
const d::vector<std::vector<const ROMol *>> & getPh4Patterns() {
 ...
 return *patterns.get();
}
```

Which makes usage more clear IMO

Reply:
It's internal, so I don care as much btw.